File: 2023/day1.py
import re


def parse_number(num: str):
    digit_names = {
        'one': '1', 'two': '2', 'three': '3', 'four': '4', 'five': '5',
        'six': '6', 'seven': '7', 'eight': '8', 'nine': '9'
    }
    if num in digit_names:
        return digit_names[num]
    return num


if __name__ == "__main__":
    with open('day1.data', 'r') as file:
        data = file.read().split('\n')

    solution_part1, solution_part2 = 0, 0
    for string in data:
        if string:
            digits = re.findall('\d', string)
            digits_and_words = re.findall('(?=(\d|one|two|three|four|five|six|seven|eight|nine))', string)
            solution_part1 += int(digits[0] + digits[-1])
            solution_part2 += int(parse_number(digits_and_words[0]) + parse_number(digits_and_words[-1]))

    print('Solution Part 1: %d' % solution_part1)
    print('Solution Part 2: %d' % solution_part2)

File: 2023-rust/src/solutions/day1.rs
use crate::utils::load_data;
use regex::Regex;

///!
///! # WARNING
///! This part of the code was originally written in Python and was just moved over to Rust.
///! Code quality might not be optimal
///!


fn parse_number(num: &str) -> String {
    let digit_names = [
        ("one", "1"), ("two", "2"), ("three", "3"), ("four", "4"), ("five", "5"),
        ("six", "6"), ("seven", "7"), ("eight", "8"), ("nine", "9")
    ];

    for &(name, digit) in digit_names.iter() {
        if num == name {
            return digit.to_string();
        }
    }

    num.to_string()
}

pub fn run() -> (i64, i64) {
    let data = load_data(1);
    let data = data.split('\n');
    let mut part1 = 0;
    let mut part2 = 0;
    let digit_regex = Regex::new("\\d").unwrap();
    let number_word_regex = Regex::new(
        "(\\d|one|two|three|four|five|six|seven|eight|nine)"
    ).unwrap();
    let number_word_regex_reverse = Regex::new(
        "(\\d|eno|owt|eerht|ruof|evif|xis|neves|thgie|enin)"
    ).unwrap();

    for datum in data {
        if !datum.is_empty() {
            let first = digit_regex.find_iter(datum).next().unwrap();
            let last = digit_regex.find_iter(datum).last().unwrap();
            let mut s = String::from(datum.get(first.start()..first.end()).unwrap());
            s.push_str(datum.get(last.start()..last.end()).unwrap());
            part1 += s.parse::<i64>().unwrap();
            let first = number_word_regex.find_iter(datum).next().unwrap();
            let rev = datum.chars().rev().collect::<String>();
            let last = number_word_regex_reverse.find_iter(rev.as_str()).next().unwrap();
            let mut s = parse_number(datum.get(first.start()..first.end()).unwrap());
            let combined: String = rev.get(last.start()..last.end()).unwrap().chars().rev().collect();
            s.push_str(parse_number(combined.as_str()).as_str());
            if let Ok(x) = s.parse::<i64>() {
                part2 += x;
            }
        }
    }
    (part1, part2)
}